src/audio/chain_dma.c

```diff
@@ -0,0 +1,587 @@
+
+	ret = chain_task_init(dev, host_dma_id, link_dma_id, fifo_size);
+	if (ret)
+		goto error_cd;
+
+	return dev;
+
+error_cd:
+	rfree(cd);
+error:
+	rfree(dev);
+	return NULL;
+}
+
+int chain_task_start(struct comp_dev *dev)
+{
+	struct comp_driver_list *drivers = comp_drivers_get();
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	int ret;
+
+	comp_info(dev, "chain_task_start(), host_dma_id = 0x%08x", cd->host_connector_node_id.dw);
+
+	key = k_spin_lock(&drivers->lock);
+	switch (cd->chain_task.state) {
+	case SOF_TASK_STATE_QUEUED:
+		k_spin_unlock(&drivers->lock, key);
+		return 0;
+	case SOF_TASK_STATE_COMPLETED:
+		break;
+	case SOF_TASK_STATE_INIT:
+		break;
+	default:
+		comp_err(dev, "chain_task_start(), bad state transition");
+		goto error;
+	}
+	cd->under_over_run_notification_sent = false;
+
+	if (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) {
+		ret = chain_host_start(dev);
+		if (ret)
+			goto error;
+		ret = chain_link_start(dev);
+		if (ret) {
+			chain_host_stop(dev);
+			goto error;
+		}
+	} else {
+		ret = chain_link_start(dev);
+		if (ret)
+			goto error;
+		ret = chain_host_start(dev);
+		if (ret) {
+			chain_link_stop(dev);
+			goto error;
+		}
+	}
+
+	pm_policy_state_lock_get(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+
+	ret = schedule_task(&cd->chain_task, 0, 0);
+	if (ret)
+		goto error;
+
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+	k_spin_unlock(&drivers->lock, key);
+
+	return 0;
+
+error:
+	k_spin_unlock(&drivers->lock, key);
+	return -EINVAL;
+}
+
+int chain_task_pause(struct comp_dev *dev)
+{
+	struct comp_driver_list *drivers = comp_drivers_get();
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	int ret;
+
+	if (cd->chain_task.state == SOF_TASK_STATE_FREE)
+		return 0;
+
+	key = k_spin_lock(&drivers->lock);
+	cd->first_data_received = false;
+	if (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) {
+		ret = chain_host_stop(dev);
+		if (ret < 0) {
+			chain_link_stop(dev);
+			goto error;
+		}
+		ret = chain_link_stop(dev);
+		if (ret < 0)
+			goto error;
+	} else {
+		ret = chain_link_stop(dev);
+		if (ret < 0) {
+			chain_host_stop(dev);
+			goto error;
+		}
+		ret = chain_host_stop(dev);
+		if (ret < 0)
+			goto error;
+	}
+
+	cd->chain_task.state = SOF_TASK_STATE_COMPLETED;
+	schedule_task_free(&cd->chain_task);
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+	k_spin_unlock(&drivers->lock, key);
+
+	return 0;
+
+error:
+	k_spin_unlock(&drivers->lock, key);
+	return -EINVAL;
+}
+
+int chain_task_trigger(struct comp_dev *dev, int cmd)
+{
+	switch (cmd) {
+	case COMP_TRIGGER_START:
+		return chain_task_start(dev);
+	case COMP_TRIGGER_PAUSE:
+		return chain_task_pause(dev);
+	default:
+		return -EINVAL;
+	}
+}
+
+void chain_task_free(struct comp_dev *dev)
+{
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+
+	chain_release(dev);
+	rfree(cd);
+	rfree(dev);
+}
+
+static enum task_state chain_task_run(void *data)
```

Comment:
fwiw, this is quite a complex function - it would benefit from more inline comments to help everyone follow.